values/decoding/fs_bool_values.go

```diff
@@ -38,13 +39,30 @@ func (v *fsBasedBoolValues) Iter() (iterator.ForwardBoolIterator, error) {
 	return newBoolIteratorFromMeta(v.metaProto, v.encodedValues)
 }
 
-// TODO(xichen): Filter implementation should take advantage of the metadata
-// to do more intelligent filtering, e.g., checking if the value is within the
-// value range.
 func (v *fsBasedBoolValues) Filter(
 	op filter.Op,
 	filterValue *field.ValueUnion,
 ) (iterator.PositionIterator, error) {
+	var (
+		numTrue  = v.Metadata().NumTrues
+		numFalse = v.Metadata().NumFalses
+	)
+	switch op {
```

Comment:
Probably safer to do a nil check `filterValue` as well as checking `filterValue.Type == field.BoolType` first before the `switch`, and return an error early otherwise.